### ISSUE TITLE
Projection: issues with multi-column foreign keys

### ISSUE THREAD

**Issue by elBranchi** [dotnet/efcore#16669]
Ok, the situation is as follows:
We are working on a new aplication that has to deal with a database with an old design.

This means that many tables have multi column primary keys, and it has the effect that there are also multi column foreign keys on other tables.

The problem comes when there are multiple relations and one (or more) of the columns is shared between the relations.

So, when querying the database and applying explicit projection (select + new {...}) or mapping projection (e.g. ProjectTo methon in AutoMapper) we get a series of undesired behaviours
when the shared column has data but not all relations exist:

* InvalidOperationException when Nullable objects that are null because the relation really does not exist.
* Properties with objects with all its properties set to null when a null value was expected instead.

On the other side, when no projection is used, by including all required tables/relation and doing the mapping on a later stage, the problem does not happen.

The problem seems to be that when checking the multi column FK, only a all-columns-are-null defines the non-existance of relation.

Is there any way provided by EFC to tune this behaviour?

I have created a test project on github, that tries to show the problem, it uses the inMemory database and sql server (localdb).
The project can be found on https://github.com/elBranchi/EFCore-Issue-Test , on master branch EFC 2.2.6 is in use, and branch efc3-preview, well you can guess :)

I have tried also usign EF Core 3 preview (3.0.0-preview6.19304.10), it seems that there are problems, too
but I haven't played around enough with 3 to sort out which parts might be usage errors or the expected result.

Also it seems that it might be related to issue #14321

### Further technical details
EF Core version:2.2.1/2.2.6
Database Provider: Microsoft.EntityFrameworkCore.SqlServer / Microsoft.EntityFrameworkCore.InMemory

**Comment by ajcvickers** [comment]
Triage: assigning this to @roji as an entity equality issue. Note that an FK should be considered null if any component is null.

**Comment by roji** [comment]
After switching to latest nightly (3.0.0-preview8.19379.1), all tests pass except for NoContactOrBothExplicitProjectionQueryTest, which fails with "Unexpected exception System.InvalidOperationException: EF.Property called with wrong property name", because of [this projection](https://github.com/elBranchi/EFCore-Issue-Test/blob/efc3-preview/EFC.Issues.Test/SharedMultiColumnFKTest.cs#L334). @smitpatel does this look familiar?

Note: none of the tests do entity equality - I see only explicit key comparison (`orderIds.Contains(od.ORDER_ID)`)

**Comment by roji** [comment]
Probably unrelated to this issue - I'm a bit unclear on exactly how nulls work with composite keys. The current logic in the entity equality rewriter - which was inherited from 2.2 - is that a null composite key is all-null. So when comparing a composite key to null, we only check the first column.

In other words, at least in entity equality we don't support having a composite key where *some* columns are null, but not all. Is this something that EF Core supports/allows?

**Comment by ajcvickers** [comment]
@roji A composite key should be considered null if any of its components are null. (Doing otherwise gets messy fast due to differences in null semantics between the database and the change tracker.)

The behavior you describe is a bug.

**Comment by roji** [comment]
@ajcvickers to be sure I understand - does EF Core itself ever generate composite keys where one column is null, but not all? If it doesn't, then the current EE behavior would be correct for EF data, but not for data inserted outside of EF. Then again, if such data exists in the database, treating all composite keys with any null value as null would also likely break in many ways...

We can discuss in triage.

**Comment by ajcvickers** [comment]
We don't support null PKs, so no part of a PK can be null.
We currently don't support null AKs, but we have an issue to support it in the future. When we do, the key will be null if any part of it is null.
We fully support nullable FKs, and if any part of the FK is null, then the key is null.

**Comment by roji** [comment]
OK. Since the entity equality logic applies only the PKs, I'm not sure there's an actual bug there at the moment - entity equality always generates comparisons against the PK properties. So unless we want to consider externally-inserted data, we should be OK for now.

**Comment by ajcvickers** [comment]
@roji Sounds good.

**Comment by smitpatel** [comment]
@roji - Duplicate of #16724

**Comment by roji** [comment]
@elBranchi @ajcvickers so it that all the provided tests will pass as expected for 3.0.

**Comment by ajcvickers** [comment]
@roji Depends whether they are valid scenarios (from an EF perspective) and also on the priority of any corner cases there.

**Comment by roji** [comment]
@ajcvickers I meant to write that everything is actually OK here - all of @elBranchi's issues have already been resolved in the latest bits, except for one test which depends on #16724, which @smitpatel is planning to fix anyway.

**Comment by elBranchi** [comment]
@ajcvickers @roji It's great to read your comments! :)

Regarding the validity of the scenarios, keep in mind that in my case we are trying to use EFC to access database that is used mainly by another application that is nearly 20 years old and built in VB6, so our usage of EFC might not fit the typical case.

**Comment by roji** [comment]
OK. So I think we can close this as a duplicate of #16724, the issues don't seem to warrant looking into for 2.2 (@ajcvickers let me know if you think differently).
